File: viewer/src/main/scala/farjs/viewer/ViewerInput.scala
package farjs.viewer

import scommons.react._
import scommons.react.blessed._
import scommons.react.hooks._

import scala.scalajs.js

object ViewerInput extends FunctionComponent[ViewerInputProps] {

  protected def render(compProps: Props): ReactElement = {
    val propsRef = useRef[ViewerInputProps](null)
    val props = compProps.plain
    val inputEl = props.inputRef.current.asInstanceOf[BlessedElement]
    propsRef.current = props
    
    useLayoutEffect({ () =>
      if (inputEl != null) {
        val keyListener: js.Function2[js.Object, KeyboardKey, Unit] = { (_, key) =>
          propsRef.current.onKeypress(key.full)
        }
        val wheelupListener: js.Function1[MouseData, Unit] = { _ =>
          propsRef.current.onWheel(true)
        }
        val wheeldownListener: js.Function1[MouseData, Unit] = { _ =>
          propsRef.current.onWheel(false)
        }

        inputEl.on("keypress", keyListener)
        inputEl.on("wheelup", wheelupListener)
        inputEl.on("wheeldown", wheeldownListener)

        val cleanup: js.Function0[Unit] = { () =>
          inputEl.off("keypress", keyListener)
          inputEl.off("wheelup", wheelupListener)
          inputEl.off("wheeldown", wheeldownListener)
        }
        cleanup
      }
      else ()
    }, List(inputEl))
    
    if (js.isUndefined(compProps.children)) null
    else compProps.children
  }
}

File: viewer/ViewerInput.mjs
/**
 * @typedef {import("@farjs/blessed").Widgets.BlessedElement} BlessedElement
 * @typedef {import("@farjs/blessed").Widgets.Events.IKeyEventArg} IKeyEventArg
 */
import React, { useLayoutEffect, useRef } from "react";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly inputRef: React.MutableRefObject<BlessedElement | null>;
 *  onWheel(up: boolean): void;
 *  onKeypress(keyFull: string): void;
 * }} ViewerInputProps
 */

/**
 * @param {React.PropsWithChildren<ViewerInputProps>} props
 * @returns {React.ReactElement | null}
 */
const ViewerInput = (props) => {
  const propsRef = useRef(props);
  propsRef.current = props;
  const inputEl = props.inputRef.current;

  useLayoutEffect(() => {
    if (inputEl) {
      /** @type {(ch: any, key: IKeyEventArg) => void} */
      const keyListener = (_, key) => {
        propsRef.current.onKeypress(key.full);
      };
      const wheelupListener = () => {
        propsRef.current.onWheel(true);
      };
      const wheeldownListener = () => {
        propsRef.current.onWheel(false);
      };

      inputEl.on("keypress", keyListener);
      inputEl.on("wheelup", wheelupListener);
      inputEl.on("wheeldown", wheeldownListener);

      return () => {
        inputEl.off("keypress", keyListener);
        inputEl.off("wheelup", wheelupListener);
        inputEl.off("wheeldown", wheeldownListener);
      };
    }

    return undefined;
  }, [inputEl]);

  return props.children ? h(React.Fragment, null, props.children) : null;
};

ViewerInput.displayName = "ViewerInput";

export default ViewerInput;
